### PR TITLE
Doc for new flatten

### LINKED ISSUE #1516: Polymorphism does not support --skip-flatten
## Problem statement
When running different fixtures with or without the `--skip-flatten` option, I found out that a good number of testcases do not support this expansion of spec.

Although the very simple `fixtures/codegen/billforward.discriminators.yml`  works, more complex test cases fail to generate code that compiles:
- `fixtures/codegen/todolist.discriminators.yml`
- test case from #606 
- test case from #607 
- test case from #1336 
- test case from #842 

I believe this behavior is more or less by design. Shouldn't we document it somewhere?

## Swagger specification
See above

## Steps to reproduce
Use:
`swagger generate model --spec=xxx --skip-flatten`

## Environment
swagger version: master
go version: 1.9.1
OS:   ubunt 14

### PR DESCRIPTION
**Pull request by fredbi**
Documents PR #1588 and issue #1572 

Fixes #1516  (doc shortcoming)

### CODE REVIEW

File: docs/use/transform.md [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+or
+
+`swagger flatten --with-expand` {spec}`
+
+or with codegen commands:
+
+`swagger generate [model|server|client|operation|...] --spec={spec} --with-expand`
+
+> **NOTE**: codegen may fail in some situations with spec expansion:
+> - polymorphism: the original intent of the `$ref` to a discriminated type(i.e. _extends type..._) is lost with expansion
```
Comment (gregmarr) [inline docs/use/transform.md]: I'm thinking "to be a discriminated type" or "as a descriminated type" would be better here (and a space before the open paren).

File: docs/use/transform.md [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+> - duplicate names: expansion of `$ref` intended for type reuse may produce duplicate type identifiers
+> - too many generated files: expansion of large specs may lead to many independent anonymous structures
+
+### Minimal flattening
+
+This transforms makes complex JSON `$ref` amenable to analysis and code generation.
```
Comment (gregmarr) [inline docs/use/transform.md]: "This transform makes" or "These transforms make"

File: docs/use/transform.md [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+Minimal flattening attempts to minimally distorts the original specification intent, while ensuring workable codegen.
```
Comment (gregmarr) [inline docs/use/transform.md]: "attempts to minimally distort the"

File: docs/use/transform.md [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+`swagger flatten --with-flatten:minimal {spec}`
+
+This is the default option for codegen commands:
+
+`swagger generate [model|server|client|operation|...] --spec={spec}`
+
+> **NOTE**: `$ref` bundling / pointer resolving may in some cases produce duplicate names.
+> The flattener tries to resolve duplicates whenever possible (such as when a child identifier duplicates its parent).
+> When such resolution is not possible, an "OAIGen" suffix is added to the definition name (a warning is issued).
+> You may use the `x-go-name` extension here to set a more suitable name to the generated type.
```
Comment (gregmarr) [inline docs/use/transform.md]: "a more suitable name for the generated type"?

**Comment by fredbi** [comment]
ok thank you @gregmarr  for your corrections.

File: docs/use/transform.md [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+This set of features is essentially provided by the `github.com/go-openapi/analysis` package.
+Feel free to contribute new features to this repo.
+
+Currently, here is a todo list of improvements planned to the spec preprocessing feature:
+
+- in full flatten mode, more options to control how allOf are handled
+- in full flatten mode propagating `x-go-name` (as prefix) to newly created definitions: 
+creating new definitions introduces some naming conventions.
+If a x-go-name has been set in the original structure, new names are no consistent with their container...
```
[the remark this answers is not in the document]
Comment (fredbi) [reply docs/use/transform.md]: done

[the remark this answers is not in the document]
Comment (fredbi) [reply docs/use/transform.md]: done

File: docs/use/transform.md [hunk 6]
```diff
@@ -0,0 +1,139 @@
+# Transforming specifications
+
+The `swagger` toolkit allows some transformation to be carried on with a specification.
```
[the remark this answers is not in the document]
Comment (fredbi) [reply docs/use/transform.md]: done

File: docs/use/transform.md (hunk 1 above)
Comment (fredbi) [reply docs/use/transform.md]: done

File: docs/use/transform.md (hunk 2 above)
Comment (fredbi) [reply docs/use/transform.md]: done

File: docs/use/transform.md (hunk 3 above)
Comment (fredbi) [reply docs/use/transform.md]: done

File: docs/use/transform.md (hunk 4 above)
Comment (fredbi) [reply docs/use/transform.md]: done

**Pull request merged**